### PR TITLE
Telemetry Job

### PR DESCRIPTION
**Pull request by shaun-nx**
### Proposed changes

This PR adds the base telemetry job that will be responsible for collecting telemetry data for each deployment of the Ingress Controller.

The current implementation of the Job writes collected, fake data to a dummy exporter.
 
Tasks to gather and report specific telemetry data points are outlined here: https://github.com/nginxinc/kubernetes-ingress/issues/4877

**Tests**
- [x] Ensure only the leader pod reports data:
```
I0124 15:43:17.788228       1 leaderelection.go:260] successfully acquired lease nginx-ingress/nginx-ingress-leader-election
I0124 15:43:17.789698       1 leader.go:60] started leading
I0124 15:43:17.789842       1 leader.go:77] updating VirtualServer and VirtualServerRoutes status
I0124 15:43:17.791000       1 telemetry.go:104] Collecting telemetry data
I0124 15:43:17.791729       1 telemetry.go:113] Exported telemetry data

```
- [x] Ensure new leader pod start reporting if leader is lost
- [x] Ensure telemetry data is not reported when `enable-telemetry-reporting` flag is set to `false`
- [x] Ensure telemetry is reported at the period that it is configured at
- [x] Update Helm values to add new CLI arguments
- [x] Confirm telemetry is not reported when `enable-telemetry-reporting` it set to `false` in Helm values
- [x] Update Helm install documentation

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

**Comment by ADubhlaoich** [comment]
Should this be mentioned somewhere in user-facing documentation?

**Comment by shaun-nx** [comment]
> Should this be mentioned somewhere in user-facing documentation?

There will be documentation added as part of the overall implementation.
This specific piece of work is just to create the base "dummy" job. It currently doesn't actually export any data.
It still needs to be fleshed out, but the issue is tracked here: https://github.com/nginxinc/kubernetes-ingress/issues/4898

File: internal/telemetry/exporter_test.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+func TestExportData(t *testing.T) {
+	t.Parallel()
+
+	exporter := NewLogExporter()
+
+	err := exporter.Export(context.Background(), Data{})
+	if err != nil {
+		t.Fatalf("Expeceted no error, but got %s", err.Error())
```
Comment (pdabelf5) [inline internal/telemetry/exporter_test.go:15]:
```suggestion
		t.Fatalf("Expected no error, but got %s", err.Error())
```

**Comment by vepatel** [comment]
@jjngx maybe we can implement a something like https://github.com/nginxinc/nginx-gateway-fabric/blob/9f4248e0d629f7953fde17f4f03bd95034306f7f/internal/mode/static/telemetry/telemetryfakes/fake_exporter.go for completeness in unit testing

File: cmd/nginx-ingress/main.go [hunk 2]
```diff
@@ -790,9 +794,12 @@ func processConfigMaps(kubeClient *kubernetes.Clientset, cfgParams *configs.Conf
...
 func updateSelfWithVersionInfo(kubeClient *kubernetes.Clientset, version, nginxVersion, appProtectVersion string, maxRetries int, waitTime time.Duration) {
-	nginxVer := strings.TrimSuffix(strings.Split(nginxVersion, "/")[1], "\n")
-	replacer := strings.NewReplacer(" ", "-", "(", "", ")", "")
-	nginxVer = replacer.Replace(nginxVer)
+	var nginxVer string
+	if nginxVersion != "" {
+		nginxVer = strings.TrimSuffix(strings.Split(nginxVersion, "/")[1], "\n")
+		replacer := strings.NewReplacer(" ", "-", "(", "", ")", "")
+		nginxVer = replacer.Replace(nginxVer)
+	}
```
Comment (pdabelf5) [inline cmd/nginx-ingress/main.go:802]: How is this change related to the telemetry work?

Comment (jjngx) [reply cmd/nginx-ingress/main.go]: This was addressed in a separate PR.

File: cmd/nginx-ingress/flags.go [hunk 3]
```diff
@@ -201,6 +203,10 @@ var (
 
 	enableDynamicSSLReload = flag.Bool(dynamicSSLReloadParam, true, "Enable reloading of SSL Certificates without restarting the NGINX process.")
 
+	enableTelemetryReporting = flag.Bool("enable-telemetry-reporting", true, "Enable gathering and reporting of product related telemetry.")
```
Comment (oseoin) [inline cmd/nginx-ingress/flags.go:206]: Do we have a task for controlling this via helm or should it be added as part of this PR?

Comment (jjngx) [reply cmd/nginx-ingress/flags.go:206]: We don't have helm-related tasks atm. We would need to add this to Helm before releasing v3.5.

Comment (shaun-nx) [reply cmd/nginx-ingress/flags.go:206]: Thanks @oseoin nice catch. I'll make this change as part of this PR.

**Review by j1m-ryan — APPROVED** [review]

File: cmd/nginx-ingress/flags.go (hunk 3 above)
Comment (jjngx) [reply cmd/nginx-ingress/flags.go:206]: Task is also added to the list.

**Review by jjngx — APPROVED** [review]

File: internal/nginx/fake_manager.go [hunk 4]
```diff
@@ -103,7 +103,7 @@ func (fm *FakeManager) CreateDHParam(_ string) (string, error) {
 // Version provides a fake implementation of Version.
 func (*FakeManager) Version() Version {
 	glog.V(3).Info("Printing nginx version")
-	return Version{}
+	return NewVersion("nginx version: nginx/1.25.3 (nginx-plus-r31)")
```
Comment (vepatel) [inline internal/nginx/fake_manager.go:106]: is this change related?

Comment (jjngx) [reply internal/nginx/fake_manager.go:106]: change related

**Review by jjngx — CHANGES_REQUESTED** [review]
We need consensus on how telemetry job time period is handled - as string (that comes at the moment from flags) or as `time.Duration`. If the latter, then unit tests would need to be updated.

**Review by vepatel — APPROVED** [review]
